spynnaker/pyNN/models/common/neuron_recorder.py

```diff
@@ -1,58 +1,520 @@
+                  "sampling_intervals for {} on one polulation. " \
+                .format(variable)
+            raise fec_excceptions.ConfigurationException(msg)
+        # else rate not changed so no action
+
+        if indexes is None:
+            # previous recording indexes does not matter as now all (None)
+            self._indexes[variable] = None
+        else:
+            self.check_indexes(indexes)
+            if self._indexes[variable] is None:
+                # just use the new indexes
+                self._indexes[variable] = indexes
+            else:
+                # merge the two indexes
+                self._indexes[variable] = \
+                    list(set(self._indexes[variable] + indexes))
+                self._indexes[variable].sort()
+
+    def set_recording(self, variable, new_state, sampling_interval=None,
+                      indexes=None):
         if variable == "all":
-            for key in self._record.keys():
-                self._record[key] = new_state
-        elif variable in self._record:
-            self._record[variable] = new_state
+            for key in self._sampling_rates.keys():
+                self.set_recording(key, new_state, sampling_interval, indexes)
+        elif variable in self._sampling_rates:
+            if new_state:
+                self._turn_on_recording(variable, sampling_interval, indexes)
+            else:
+                self._turn_off_recording(variable, sampling_interval, indexes)
         else:
             msg = "Variable {} is not supported ".format(variable)
             raise fec_excceptions.ConfigurationException(msg)
 
-    def get_sdram_usage_in_bytes(self, variable, n_neurons,
-                                 n_machine_time_steps):
-        if self.is_recording(variable):
-            return recording_utils.get_recording_region_size_in_bytes(
-                n_machine_time_steps,  self.N_BYTES_PER_NEURON * n_neurons)
+    def get_buffered_sdram_per_record(self, variable, vertex_slice):
+        """
+        Returns the sdram used per record
+
+        :param variable:
+        :param vertex_slice:
+        :return:
+        """
+        n_neurons = self._count_recording_per_slice(variable, vertex_slice)
+        if n_neurons == 0:
+            return 0
+        if variable == SPIKES:
+            if n_neurons < vertex_slice.n_atoms:
+                # Indexing is used rather than gating to dettermine recording
+                # Non recoding neurons write to an extra slot
+                n_neurons += 1
+            out_spike_words = int(math.ceil(n_neurons / 32.0))
+            out_spike_bytes = out_spike_words * self.N_BYTES_PER_WORD
+            return self.N_BYTES_FOR_TIMESTAMP + out_spike_bytes
         else:
+            return self.N_BYTES_FOR_TIMESTAMP + \
+                        n_neurons * self.N_BYTES_PER_VALUE
+
+    def get_buffered_sdram_per_timestep(self, variable, vertex_slice):
```

Comment:
It might make more sense for this to be get_n_timesteps_in_buffer_space (as this is the actual call that gets made to the vertices where this is used).  This would then avoid having to return a fractional value when only some timesteps are being recorded.  Places that would have to be changed to make this work are (these are also tagged in the review):
1. You wouldn't need get_sampling_overflow_sdram below
1. You wouldn't need _get_buffered_sdram_per_timestep in abstract_population_vertex.
1. PopulationMachineVertex would need to be passed the neuron recorder.  
1. PopulationMachineVertex would then call neuron_recorder.get_n_timesteps_in_buffer_space from its own version of that method.